Add test for inserting anchor without tx

remix7531 opened the pull request:
### Description

This adds a test case described in [issue958](https://github.com/bitcoindevkit/bdk/issues/958)

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

evanlinjin commented on crates/chain/tests/test_tx_graph.rs:
```diff
@@ -880,3 +881,57 @@ fn test_additions_last_seen_append() {
+    let mut graph = TxGraph::<BlockId>::default();
+    let txs = [common::new_tx(0), common::new_tx(1)];
+    let chain: LocalChain = (0..=5)
+        .map(|ht| (ht, BlockHash::hash(format!("Block Hash {}", ht).as_bytes())))
```
We have a `h!("data")` that can be used.

remix7531 replied:
`h!` only allows `literals`. Going to change to `expr` to be more general

evanlinjin commented on crates/chain/tests/test_tx_graph.rs:
```diff
@@ -880,3 +881,57 @@ fn test_additions_last_seen_append() {
+        }
+    );
+
+    graph.apply_additions(additions);
```
This is interesting. You are re-adding things that are already added. I don't think this should be here, but it would be something interesting to test in a separate test.

The result of adding things that are already added must be no changes being made.

remix7531 replied:
No. I am using [`insert_anchor_preview()`](https://bitcoindevkit.org/docs-rs/bdk/nightly/latest/bdk_chain/tx_graph/struct.TxGraph.html#method.insert_anchor_preview) to generate additions.

evanlinjin reviewed (changes requested):
Thanks for the work so far. Remember that we also want to test recovering the appended additions into a new graph. Refer to https://github.com/bitcoindevkit/bdk/issues/958#issuecomment-1532394373

evanlinjin commented on crates/chain/tests/test_tx_graph.rs:
```diff
@@ -822,3 +822,71 @@ fn test_additions_last_seen_append() {
+            .get_chain_position(&chain, chain.tip().unwrap(), txs[0].txid())
+            .unwrap(),
+        ObservedAs::Confirmed(&con_hei_anchor)
+    );
```
Thanks for this, there is a `TxGraph::all_anchors` method now.

evanlinjin commented on crates/chain/tests/test_tx_graph.rs:
```diff
@@ -822,3 +822,71 @@ fn test_additions_last_seen_append() {
+        .map(|i| (i, h!(i.to_string())))
+        .collect::<BTreeMap<u32, BlockHash>>()
+        .into();
+    let con_hei_anchor = ConfirmationHeightAnchor {
```
Nit: just name this variable `anchor`.

evanlinjin commented on crates/chain/tests/test_tx_graph.rs:
```diff
@@ -822,3 +822,71 @@ fn test_additions_last_seen_append() {
+
+    // graph should be recoverable from additions
+    graph_recovered.apply_additions(additions2);
+    graph_recovered.apply_additions(additions3);
```
This is cool, but I intended we test "merging" the additions first.

```rust
let mut acc_additions = Additions::default();

// maybe have this as the body of a loop over txs
{
    let additions = graph.insert_tx(tx);
    assert_eq!(additions, todo!());
    acc_additions.append(additions);
}

// recover graph with accumulated additions
let mut recovered_graph = TxGraph::default();
recovered_graph.apply_additions(acc_additions);
assert_eq!(recovered_graph, graph, "accumulated additions from 'graph' should recover the same 'graph'");
```

evanlinjin commented on crates/chain/tests/test_tx_graph.rs:
```diff
@@ -822,3 +822,71 @@ fn test_additions_last_seen_append() {
+        },
+        confirmation_height: 0,
+    };
+    let additions = graph.insert_anchor(txs[0].txid(), con_hei_anchor);
```
Could we test inserting multiple anchors as well? (as per spec)

evanlinjin commented on crates/chain/tests/test_tx_graph.rs:
```diff
@@ -822,3 +822,71 @@ fn test_additions_last_seen_append() {
     }
 }
+
+#[test]
```
I think it will be nice to have a comment summarizing what is tested here.

evanlinjin reviewed (changes requested):
Good work so far. Few notes.

1. Ensure CI passes.
2. Ensure GitHub can verify your commits (Maybe you need to give your PGP or SSH public key to GitHub?).
3. Please rebase your commits into one.
4. Use loops where you can (for example if you are inserting multiple anchors or multiple txs).

Commit:
Adds test inserting anchor without tx

Test inserting anchors without the corresponding transaction
and testing the recoverability of the transaction graph

Partially fixes "More tests for IndexedTxGraph and TxGraph #958"
https://github.com/bitcoindevkit/bdk/issues/958

danielabrozzoni reviewed (commented):
A few improvements can be made:
- Fix the warnings `cargo clippy` is throwing
- Use a more descriptive commit message. Instead of "tests for issue958", describe which test you're adding in the title, and add "Partially fixes #958" in the description. In general, it should be possible to understand what a commit is doing from the title/description without having to open github

The rest looks good to me.

evanlinjin commented on crates/chain/tests/test_tx_graph.rs:
```diff
@@ -822,3 +822,95 @@ fn test_additions_last_seen_append() {
+                (anchors[2], txs[2].txid()),
+            ]),
+            last_seen: Default::default(),
+        }
```
Nit, for future reference.

```suggestion
        Additions {
            anchors: BTreeSet::from([
                (anchors[0], txs[0].txid()),
                (anchors[1], txs[1].txid()),
                (anchors[2], txs[2].txid()),
            ]),
            ..Default::default()
        }
```

evanlinjin commented on crates/chain/tests/test_tx_graph.rs:
```diff
@@ -822,3 +822,95 @@ fn test_additions_last_seen_append() {
+                txout: Default::default(),
+                anchors: BTreeSet::default(),
+                last_seen: Default::default(),
+            }
```
Nit
```suggestion
            Additions {
                tx: BTreeSet::from([tx.clone()]),
                ..Default::default()
            }
```

evanlinjin commented on crates/chain/tests/test_tx_graph.rs:
```diff
@@ -822,3 +822,95 @@ fn test_additions_last_seen_append() {
+            ObservedAs::Unconfirmed(0)
+        );
+    }
+}
```
I don't think it makes sense to use `get_chain_position` for this test. It feels "out of scope". We want a direct way of seeing if we have the correct anchors inserted for a given tx. I would use `get_tx_node` here. This also simplifies the test as we don't need a `LocalChain`.

Note that `get_tx_node` only returns `Some` for full transactions.

evanlinjin replied:
I think checking both `get_tx_node` and `all_anchors` will be good here.

evanlinjin commented on crates/chain/tests/test_tx_graph.rs:
```diff
@@ -822,3 +822,95 @@ fn test_additions_last_seen_append() {
+    // Check whether Anchors are part of TxGraph
+    for i in 0..3 {
+        assert!(graph.all_anchors().contains(&(anchors[i], txs[i].txid())));
+    }
```
This is another way of writing this, but I guess it's a matter of taste. Not mandatory (no pun intended).

```suggestion
    assert_eq!(
        graph.all_anchors(),
        &anchors
            .iter()
            .cloned()
            .zip(txs.iter().map(Transaction::txid))
            .take(3)
            .collect()
    );
```

evanlinjin reviewed (changes requested):
Thank you again, sorry I think it will be better to reduce the scope of this test.

Commit:
Simplify the test. Remove LocalChain

evanlinjin commented on crates/chain/tests/test_tx_graph.rs:
```diff
@@ -822,3 +822,90 @@ fn test_additions_last_seen_append() {
+        assert!(graph.get_tx_node(tx.txid()).unwrap().anchors.is_empty());
+    }
+
+    assert!(false)
```
Woops!

Commit:
Fix assert!(false) and clippy warnings

notmandatory commented:
Since it looks like still a bit of work today I'm moving this out of the alpha.1 milestone.

remix7531 committed:
Merge branch 'master' into issue958

remix7531 commented:
I do not understand what is missing here. The workflows failed because of [this](https://github.com/bitcoindevkit/bdk/actions/runs/5177377132/jobs/9624816846#step:5:13) 
`error: package log v0.4.19 cannot be built because it requires rustc 1.60.0 or newer, while the currently active rustc version is 1.57.0`

Just tested the workflows locally and it seams to work now.

danielabrozzoni commented:
Hey @remix7531, sorry for the late response! I think you encountered a problem with our CI, that's why it would work locally but fail here. Can you try rebasing?

oleonardolima commented:
Is this something still needed for 1.0.0-beta milestone ? (cc @notmandatory @ValuedMammal)

I got a bit confused because I noticed that the #958 issue already has been closed but without the tests covered by this PR.

ValuedMammal commented:
I think it wouldn't hurt to have a test similar to this

notmandatory commented:
This testing could help with verifying sqlite schema question in #1712

evanlinjin commented:
Tests for inserting anchor without tx already exist: https://github.com/bitcoindevkit/bdk/blob/f6c1c61e7611966e20a88aace932a08aabd543f2/crates/chain/tests/test_tx_graph.rs#L1074-L1102
